text/MateuszTokarz/opis.tex

```diff
@@ -0,0 +1,99 @@
+
+Pozostaje ograniczyć współczynnik aproksymacji algorytmu.
+
+\section*{Współczynnik aproksymacji}
+
+\begin{tw}
+Algorytm przedstawiony powyżej jest $log(n)$ aproksymacyjny.
+\end{tw}
+\begin{proof} [Dowód]
+Niech $s_1, \dots, s_m$ będzie ułożeniem słów wejściowych w kolejności ich występowania w optymalnym rozwiązaniu - jeżeli słowo występuje kilka razy, to rozważamy najbardziej lewe wystąpienie. Dzielimy słowa na następujące grupy:
+\newline
+Grupa $G_1$ zawiera $s_1, \dots, s_i$, gdzie $s_i$ to ostatnie słowo takie, że pierwsze wystąpienie $s_1$ nachodzi na pierwsze wystąpienie $s_i$ w optymalnym rozwiązaniu.
+\newline
+Grupa $G_2$ budowana jest w analogiczny sposób, z tym że zaczynamy konstrukcje od $s_{i+1}$. W ten sposób dzielimy całe wejście $s_1, \dots, s_m$ na rozłączne grupy $G_1, \dots, G_k$.
+
+Dla każdej grupy $G_i$ wyróżniamy elementy $b_i, t_i$ - pierwsze i ostatnie słowa (we wcześniej wspomnianym porządku) budujące grupę. Jako że $b_i$ nachodzi na $t_i$, to istnieje scalenie $m_i$ takie, że wszystkie słowa z grupy $G_i$ są podsłowami $m_i(b_i, t_i)$.
+Zauważmy, że 
+$$\sum_{i=1}^{k} |m_i(b_i, t_i)| \leq 2n,$$
+gdzie $n$ to długość optymalnego rozwiązania, ponieważ każda litera słowa optymalnego jest pokrywana przez maksymalnie dwie grupy.
+
+Wprowadźmy teraz pojęcie kosztu słowa w kontekście analizowanego przez nas algorytmu. Załóżmy, że słowo $s$ odpadało ze zbioru $S$ na rzecz scalenia $m(s', s'')$. 
+$$cost(s) = \frac{|m(s', s'')|}{v(m(s', s''))} \cdot |s|.$$
+
+Zauważmy, że koszt całego algorytmu (długość słowa wynikowego) jest równa nie więcej niż długość wszystkich słów, które powstały w trakcie pierwszej fazy opróżniania zbioru $S$ - elementów $s_1, \dots, s_m$. Niech $S_h = A(M_h(B_h, T_h))$ to zbiór elementów wyrzuconych $h$-tym kroku wspomnianej fazy. 
+$$COST \leq \sum_{h=1}^{r} |M_h(B_h, T_h)| = \sum_{i=h}^{r} \frac{|M_h(B_h, T_h)|}{v(M_h(B_h, T_h))} \cdot v(M_h(B_h, T_h))= $$
+$$\sum_{h=1}^{r} \sum_{s \in S_h} \frac{|M_h(B_h, T_h)|}{v(M_h(B_h, T_h))} \cdot |s| = \sum_{h=1}^{r} \sum_{s \in S_h} cost(s) = \sum_{s \in S} cost(s).$$
+\end{proof}
+
+Rozważmy teraz grupę $G_j$ i koszt, jaki płacimy za wyrzucenie jej elementów. Grupę $G_j$ dokładnie przed fazą $h$ będziemy oznaczać $G_j^{h}$.
+Sumę długości słów w zbiorze $A$ oznaczamy $||A||$.
+ 
+
+$$Cost(G_j) = \sum_{h=1}^{k} \sum_{s \in (G_j^{h+1} \backslash G_j^{h})} cost(s) = \sum_{h=1}^{k} \frac{|M_h(B_h, T_h))|}{v(M_h(B_h, T_h))} \cdot (||G_j^{h+1}|| - ||G_j^{h}||).$$
+Jako że nasz algorytm wybierał scalenia $M_h(B_h, T_h)$ realizujące minimalne wartości, to
+$$\sum_{h=1}^{k} \frac{|M_h(B_h, T_h))|}{v(M_h(B_h, T_h))} \cdot (||G_j^{h+1}|| - ||G_j^{h}||) \leq \sum_{h=1}^{k} \frac{|m_j(b_j, t_j))|}{v(m_j(b_j, t_j))} \cdot (||G_j^{h+1}|| - ||G_j^{h}||) = $$
+$$\sum_{h=1}^{k} \frac{|m_j(b_j, t_j))|}{||G_j^{h}||} \cdot (||G_j^{h+1}|| - ||G_j^{h}||) \leq |m_j(b_j, t_j))| \cdot Harm(||G_j||).$$
```

Comment:
What is Harm? I guess it's ||G_j||-th harmonic number, but frankly that's the first time I see such a notation.
Moreover, it's not entirely clear why sum_{h=1}^{k} \frac{||G_j^{h+1}|| - ||G_j^{h}||}{||G_j^{h}||} \le Harm(||G_j||) - please provide some one/two-line explanation.